runtime-transaction/src/instruction_details.rs

```diff
@@ -0,0 +1,273 @@
+use {
+    crate::{builtin_instruction_details::*, compute_budget_instruction_details::*},
+    solana_compute_budget::compute_budget_limits::*,
+    solana_sdk::{
+        instruction::CompiledInstruction,
+        pubkey::Pubkey,
+        transaction::{Result, TransactionError},
+    },
+    std::num::NonZeroU32,
+};
+
+/// Information about instructions gathered after scan over transaction;
+/// These are "raw" information that suitable for cache and reuse.
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Default, Debug)]
+pub struct InstructionDetails {
+    builtin_instruction_details: BuiltinInstructionDetails,
+    compute_budget_instruction_details: ComputeBudgetInstructionDetails,
+}
+
+impl InstructionDetails {
+    pub fn try_from<'a>(
+        instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
+    ) -> Result<Self> {
+        let mut compute_budget_instruction_details = ComputeBudgetInstructionDetails::default();
+        let mut builtin_instruction_details = BuiltinInstructionDetails::default();
+
+        for (i, (program_id, instruction)) in instructions.enumerate() {
+            compute_budget_instruction_details.process_instruction(
+                i as u8,
+                program_id,
+                instruction,
+            )?;
+            builtin_instruction_details.process_instruction(program_id, instruction)?;
```

Comment:
I find this API very strange. I would prefer if we did something like add a method to `SanitizedMessage` called something like `try_process_compute_budget_instructions` or use a `ComputeBudgetInstructionDetails::try_from_message` constructor which takes the `SanitizedMessage` as an argument.

Reply:
continue from above discussion, say if we filtered a list of interested instructions, which might be a mix of builtins, pre-compiles, compute-budgets and perhaps some other types; It's still good to iterate through the list once, pass each instruction for different processors to collect individual info. So we are operates on per-instruction basis, instead of per `SanitizedMessage` (or `SVMMessage`). Here it loosely implements Visitor pattern, where different instructions are processed by different processors. Might be better to refactor to something like this: 
```
for ix in fitlered_instructions {
    ix.accept(&mut compute_budget_instruction_details)?;
    ix.accept(&mut builtin_instruction_details)?;
    ... ...
}
```

Reply:
Yeah the visitor pattern is fine I guess but I don't like that this interface operates over individual unfiltered instructions and does a hash for each program id as @apfitzge mentioned above.

Reply:
been doing bench and profiling on iteration and hashing. Will share finding later.

Reply:
a782e33a3205352d882c65184d9cf1c459f56ef4 adds an old-school quick builtins-filter, using the bench from #2498 , it addresses the concern: it no longer wastes time to hash non-builtin program key to check against hashmap. 

<details>
<summary>bench of master</summary>

```
running 5 tests
test bench_process_compute_budget_instructions_builtins        ... bench:      16,077 ns/iter (+/- 397)
test bench_process_compute_budget_instructions_compute_budgets ... bench:      31,168 ns/iter (+/- 952)
test bench_process_compute_budget_instructions_empty           ... bench:       3,264 ns/iter (+/- 20)
test bench_process_compute_budget_instructions_mixed           ... bench:     449,149 ns/iter (+/- 6,138)
test bench_process_compute_budget_instructions_non_builtins    ... bench:      16,066 ns/iter (+/- 132)
```

</details>

<details>
<summary>bench before this commit - almost doubled time for all scenarios</summary>

```
running 5 tests
test bench_process_compute_budget_instructions_builtins        ... bench:      42,953 ns/iter (+/- 1,501)
test bench_process_compute_budget_instructions_compute_budgets ... bench:      43,980 ns/iter (+/- 820)
test bench_process_compute_budget_instructions_empty           ... bench:       6,138 ns/iter (+/- 176)
test bench_process_compute_budget_instructions_mixed           ... bench:   1,353,764 ns/iter (+/- 27,608)
test bench_process_compute_budget_instructions_non_builtins    ... bench:      35,780 ns/iter (+/- 1,129)
```

</details>

<details>
<summary>bench with this commit - non_builtins perf back t o original</summary>

```
running 5 tests
test bench_process_compute_budget_instructions_builtins        ... bench:      40,815 ns/iter (+/- 269)
test bench_process_compute_budget_instructions_compute_budgets ... bench:      44,544 ns/iter (+/- 155)
test bench_process_compute_budget_instructions_empty           ... bench:       5,770 ns/iter (+/- 49)
test bench_process_compute_budget_instructions_mixed           ... bench:     389,491 ns/iter (+/- 5,886)
test bench_process_compute_budget_instructions_non_builtins    ... bench:      15,358 ns/iter (+/- 515)
```

</details>

Reply:
But this still shows the first 2 scenarios as having much worse performance, no?

Reply:
yea, these two demonstrate the added cost of locating builtin program cost from BUILTIN_INSTRUCTION_COSTS, but to builtins only